### PR TITLE
Multi dest connector

### PR DESCRIPTION
**Pull request by xazhao**
### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [x] Add/update [unit tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions) using:
    - [ ] Correct values
    - [ ] Bad/wrong values (None, empty, wrong type, length, etc.)
    - [ ] [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html)
- [x] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected
- [ ] Do these changes include any template validations?
    - [ ] Did the newly validated properties support intrinsics prior to adding the validations? (If unsure, please review [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) before proceeding).
        - [ ] Does the pull request ensure that intrinsics remain functional with the new validations?

### Examples?

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/sam_resources.py [hunk 1]
```diff
@@ -1785,88 +1787,106 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
...
         replace_depends_on_logical_id(self.logical_id, generated_logical_ids, resource_resolver)
 
-        self._add_connector_metadata(generated_resources, original_template, source, destination)
-        if generated_resources:
-            return generated_resources
+        if list_generated_resources:
+            return list_generated_resources
 
         # Should support all profile types
         raise TypeError(f"Unknown profile policy type '{profile_type}'")
```
Comment (aahung) [inline samtranslator/model/sam_resources.py:1892]: since this variable `profile_type` was created in the `for` loop, it doesn't make sense to use it outside the loop anymore

File: samtranslator/model/sam_resources.py [hunk 2]
```diff
@@ -1785,88 +1787,106 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
...
+                )
+            if profile_type == "AWS_SNS_TOPIC_POLICY":
+                generated_resources.append(
+                    self._construct_sns_topic_policy(source, destination, profile_properties, index, multi_dest)
+                )
+            if profile_type == "AWS_LAMBDA_PERMISSION":
+                generated_resources.extend(
+                    self._construct_lambda_permission_policy(source, destination, profile_properties, index, multi_dest)
+                )
+
```
Comment (aahung) [inline samtranslator/model/sam_resources.py:1881]: here we should do a `raise TypeError(f"Unknown profile policy type '{profile_type}'")`

File: samtranslator/model/sam_resources.py (hunk 1 above)
Comment (aahung) [reply samtranslator/model/sam_resources.py:1892]: this should change to "the connector destination is empty"

**Review by aahung — COMMENTED** [review]
Overall it looks good! Just some minor error handling issues. Can we make sure those errors are correct using some transform tests?

File: samtranslator/model/sam_resources.py [hunk 3]
```diff
@@ -1785,88 +1787,106 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
...
 
-        # removing duplicate permissions
-        self.Permissions = list(set(self.Permissions))
-        profile_type, profile_properties = profile["Type"], profile["Properties"]
-        profile_permissions = profile_properties["AccessCategories"]
-        valid_permissions_combinations = profile_properties.get("ValidAccessCategories")
+        list_generated_resources: List[Resource] = []
 
-        valid_permissions_str = ", ".join(profile_permissions)
+        for index, dest in enumerate(self.Destination):
```
Comment (hoffa) [inline samtranslator/model/sam_resources.py:1797]: A large nested function is hard to read. Should be possible to separate the concerns into logical parts; adding a 1:1 connector is one thing, and multi-destination is a generalization of that.

File: samtranslator/model/sam_resources.py [hunk 4]
```diff
@@ -1923,6 +1947,8 @@ def _construct_lambda_permission_policy(
         source: ConnectorResourceReference,
         destination: ConnectorResourceReference,
         profile: ConnectorProfile,
+        index: int,
+        multi_dest: bool,
```
Comment (hoffa) [inline samtranslator/model/sam_resources.py:1951]: Could have a single variable by e.g. using `index: Optional[int]` instead.

File: samtranslator/model/sam_resources.py (hunk 3 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1797]: Synced offline. Will move the majority of for loop into a function `generate_resource`.

File: samtranslator/model/sam_resources.py (hunk 4 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1951]: the index is from the loop. So I think we still need `multi_dest` flag to decide it's a single destination or a list that only includes 1 destination.

File: samtranslator/model/sam_resources.py (hunk 1 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1892]: Good catch! Will update it.

File: samtranslator/model/sam_resources.py (hunk 2 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1881]: agree

File: samtranslator/model/sam_resources.py (hunk 4 above)
Comment (hoffa) [reply samtranslator/model/sam_resources.py:1951]: `index` is only used for multi-resource connectors right? An "index" doesn't make sense for a non-list structure anyway, so I don't see why it should be relevant when not using multi-resource.

Comment (hoffa) [reply samtranslator/model/sam_resources.py:1951]: (Also might want to make that `dest_index` for clarity. There's a good change we'll do multi-source as well.)

**Comment by xazhao** [comment]
@aahung Actually those errors can not be approached by transform test. They will be errors before.

File: samtranslator/model/sam_resources.py [hunk 5]
```diff
@@ -1770,42 +1770,41 @@ class SamConnector(SamResourceMacro):
...
             raise InvalidResourceException(
                 self.logical_id,
                 f"Unable to create connector from {source.resource_type} to {destination.resource_type}; it's not supported or the template is invalid.",
             )
 
+        profile_type, profile_properties = profile["Type"], profile["Properties"]
+
         # removing duplicate permissions
         self.Permissions = list(set(self.Permissions))
-        profile_type, profile_properties = profile["Type"], profile["Properties"]
```
Comment (hoffa) [inline samtranslator/model/sam_resources.py:1803]: nit: should avoid unrelated changes (especially for cognitively high-load PRs like this), makes reviewing harder

File: samtranslator/model/sam_resources.py [hunk 6]
```diff
@@ -1816,7 +1815,7 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
             if permission not in profile_permissions:
                 raise InvalidResourceException(
                     self.logical_id,
-                    f"Unsupported 'Permissions' provided; valid values are: {valid_permissions_str}.",
+                    f"Unsupported 'Permissions' provided for connector from {source.resource_type} to {destination.resource_type}; valid values are: {valid_permissions_str}.",
```
[the remark this answers is not in the document]
Comment (GavinZZ) [reply samtranslator/model/sam_resources.py:1818]: > Why the change?
> 
> The error message already includes the logical ID which allows looking up any information necessary. Also `resource_type` is the CloudFormation type, so it doesn't necessarily match what the customer uses.

If users are using multi dest connector, isn't it clearer to tell which source-destination pair results in this error message?

File: samtranslator/model/sam_resources.py [hunk 7]
```diff
@@ -1848,28 +1848,62 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
...
 
-        self._add_connector_metadata(generated_resources, original_template, source, destination)
-        if generated_resources:
-            return generated_resources
+        if list_generated_resources:
+            return list_generated_resources
 
-        # Should support all profile types
-        raise TypeError(f"Unknown profile policy type '{profile_type}'")
+        raise TypeError("The destination is empty")
```
Comment (GavinZZ) [inline samtranslator/model/sam_resources.py:1906]: nit: This error message doesn't seem quite right. There are other scenarios where `list_generated_resources` can be empty list.

File: samtranslator/model/sam_resources.py [hunk 8]
```diff
@@ -1888,6 +1922,8 @@ def _construct_iam_policy(
         destination: ConnectorResourceReference,
         profile: ConnectorProfile,
         resource_resolver: ResourceResolver,
+        dest_index: int,
+        multi_dest: bool,
```
Comment (GavinZZ) [inline samtranslator/model/sam_resources.py:1926]: Instead of these two variables, can we make `dest_index` as a optional? The checks below can become 
`policy_name = f"{self.logical_id}PolicyDestination{dest_index}" if dest_index != None else f"{self.logical_id}Policy"`

File: samtranslator/model/sam_resources.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1818]: Synced with Chris offline. It's clearer to have customer know which resource pair doesn't support the permission.

File: samtranslator/model/sam_resources.py [hunk 9]
```diff
@@ -1848,28 +1848,62 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
...
+        list_generated_resources: List[Resource] = []
+
+        for dest_index, dest in enumerate(self.Destination):
+            try:
+                destination = get_resource_reference(dest, resource_resolver, self.Source)
+                source = get_resource_reference(self.Source, resource_resolver, dest)
+            except ConnectorResourceError as e:
+                raise InvalidResourceException(self.logical_id, str(e))
+
+            generated_resource = self.generate_resource(source, destination, dest_index, multi_dest, resource_resolver)
```
Comment (hoffa) [inline samtranslator/model/sam_resources.py:1895]: This should be _resources_, right?

File: samtranslator/model/sam_resources.py (hunk 7 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1906]: Do you know what scenarios `list_generated_resources` can be empty list? Trying to figure out the right message here.

File: samtranslator/model/sam_resources.py (hunk 5 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1803]: will update it.

**Review by GavinZZ — COMMENTED** [review]
Generally LGTM, just some minor comments.

File: samtranslator/model/sam_resources.py (hunk 9 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1895]: yes.

Comment (GavinZZ) [reply samtranslator/model/sam_resources.py:1895]: `generated_resource` seems right here. It's passing only this specific source and destination pair.

File: samtranslator/model/sam_resources.py (hunk 7 above)
Comment (aahung) [reply samtranslator/model/sam_resources.py:1906]: I don't think this will ever happen if our profile is right. So if it happens, it should be an availability dip. We can raise something like unexpected error

File: samtranslator/model/sam_resources.py (hunk 9 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1895]: For lambda permission with `READ + WRITE`, we creates 2 lambda permissions. In other cases, we only create 1 policy

File: samtranslator/model/sam_resources.py (hunk 8 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py:1923]: We would still need a flag to identify it's a single destination or a list only contains one destination.

File: samtranslator/model/sam_resources.py (hunk 7 above)
Comment (xazhao) [reply samtranslator/model/sam_resources.py]: updates: an empty destination list will trigger this error. So I updated to `InvalidResourceException` because it's a template error.

**Review by GavinZZ — APPROVED** [review]

File: samtranslator/model/sam_resources.py [hunk 10]
```diff
@@ -1850,26 +1848,61 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
...
+                self._construct_lambda_permission_policy(
+                    source, destination, profile_properties, dest_index, multi_dest
+                )
+            )
+        else:
+            raise TypeError(f"Profile type {profile_type} is not supported")
+        return generated_resources
+
+    @cw_timer
+    def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
```
Comment (hoffa) [inline samtranslator/model/sam_resources.py:1874]: Why `# type: ignore`?

File: samtranslator/model/sam_resources.py [hunk 11]
```diff
@@ -1850,26 +1848,61 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
...
 
-        self._add_connector_metadata(generated_resources, original_template, source, destination)
-        if generated_resources:
-            return generated_resources
+        if list_generated_resources:
+            return list_generated_resources
 
-        # Should support all profile types
-        raise TypeError(f"Unknown profile policy type '{profile_type}'")
+        raise InvalidResourceException(self.logical_id, "The destination is an empty list")
```
Comment (hoffa) [inline samtranslator/model/sam_resources.py:1905]: To match the wording so far:

```suggestion
        raise InvalidResourceException(self.logical_id, "'Destination' is an empty list")
```

**Review by hoffa — APPROVED** [review]

**Review by aahung — APPROVED** [review]
:shipit:
